book/src/rfcs/RFC0007-Opinionated-Cancelation.md

```diff
@@ -0,0 +1,63 @@
+* Modify salsa queries to automatically panic when they are canceled
+* Use a distinguished value for this panic so that people can test if the panic was a result of cancellation
+
+## Motivation
+
+Salsa's database model is fundamentally like a read-write lock. There is always a single *master copy* of the database which supports writes, and any number of concurrent *snapshots* that support reads. Whenever a write to the database occurs, any queries executing in those snapshots are considered *canceled*, because their results are based on stale data. The write blocks until they complete before it actually takes effect. It is therefore advantageous for those reads to complete as quickly as possible.
+
+Cancellation in salsa is currently quite minimal. Effectively, a flag becomes true, and queries can manually check for this flag. This is easy to forget to do. Moreover, we support two modes of cancellation: you can either use `Result` values or use `panic!`. In practice, though, there isn't much point to using `Result`: you can't really "recover" from cancellation.
+
+The largest user of salsa, rust-analyzer, uses a fairly opinionated and aggressive form of cancellation:
+
+* Every query is instrumented, using salsa's various hooks, to check for cancellation before it begins.
+* If a query is canceled, then it immediately panics, using a special sentinel value.
+* Any worker threads holding a snapshot of the DB recognize this value and go back to waiting for work.
+
+We propose to make this model of cancellation the *only* model of cancellation.
+
+## User's guide
+
+When you do a write to the salsa database, that write will block until any queries running in background threads have completed. You really want those queries to complete quickly, though, because they are now operating on stale data and their results are therefore not meaningful. To expedite the process, salsa will *cancel* those queries. That means that the queries will panic as soon as they try to execute another salsa query. Those panics occur using a sentinel value that you can check for if you wish. If you have a query that contains a long loop which does not execute any intermediate queries, salsa won't be able to cancel it automatically. You may wish to check for cancelation yourself by invoking the `panic_if_canceled` method.
+
+## Reference guide
+
+The changes required to implement this RFC are as follows:
+
+* Adjust comments on `is_current_revision_canceled`.
+* Introduce a sentinel value `static SALSA_CANCELATION_TOKEN: SalsaCancelationToken = SalsaCancelationToken` that can be used with `panic_any!`
+* Introduce a `panic_if_canceled` method into the runtime which checks whether cancelation has occured and panics if so.
+    * This should probably be inline for the `if` with an outlined function to do the actual panic.
+* Modify the code for the various queries to invoke `panic_if_canceled` when they are invoked.
```

Comment:
Not only invoked, but also when validated. If you have a complex query graph which is fresh, but we don’t know it yet, validating it will might take significant time, and it’s advantageous to cancel validation midway.

We did hit “slow validation causes lag” in rust-analyzer, although *maaaybe* that was before we introduced durability to speed up validation?